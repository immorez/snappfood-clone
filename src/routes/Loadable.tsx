/* eslint-disable react/display-name */
import { ElementType, Suspense } from "react";

import LoadingScreen from "../components/LoadingScreen";

/**
 * Loadable is a higher-order component (HOC) that wraps a given component
 * with React Suspense to handle code-splitting and display a loading screen
 * while the component is being loaded.
 *
 * @param {ElementType} Component - The component to be lazily loaded.
 * @returns {Function} A function that represents the wrapped component.
 */

const Loadable = (Component: ElementType) =>
    function (props: any) {
        return (
            <Suspense fallback={<LoadingScreen />}>
                <Component {...props} />
            </Suspense>
        );
    };

export default Loadable;
