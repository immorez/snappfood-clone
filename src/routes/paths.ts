import path from "@/utils/path";

const ROOTS_MAIN = "";

export const PATH_MAIN = {
    root: ROOTS_MAIN,
    home: path(ROOTS_MAIN, "/"),
    restaurant: path(ROOTS_MAIN, "/restaurants"),
};
