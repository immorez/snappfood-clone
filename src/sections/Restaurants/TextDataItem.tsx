import React from "react";

const TextDataItem = function () {
    return <div>text data item</div>;
};

export default React.memo(TextDataItem);
