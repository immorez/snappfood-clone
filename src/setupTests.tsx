import "jest";
import "@testing-library/jest-dom";
import { RenderOptions, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "@/redux/store";
import { I18nextProvider } from "react-i18next";
import i18n from "./utils/i18n";
import {
    BrowserRouter,
    MemoryRouter,
    MemoryRouterProps,
} from "react-router-dom";
import { ReactElement } from "react";

interface AllTheProvidersProps {
    children: React.ReactNode;
    routerOptions?: MemoryRouterProps;
    useMemoryRouter?: boolean;
}

const AllTheProviders: React.FC<AllTheProvidersProps> = ({
    children,
    routerOptions,
    useMemoryRouter = false,
}) => {
    const RouterComponent = useMemoryRouter ? MemoryRouter : BrowserRouter;
    return (
        <Provider store={store}>
            <I18nextProvider i18n={i18n}>
                <RouterComponent {...routerOptions}>{children}</RouterComponent>
            </I18nextProvider>
        </Provider>
    );
};

const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, "wrapper"> & Partial<AllTheProvidersProps>,
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
