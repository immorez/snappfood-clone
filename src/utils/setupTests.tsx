import { RenderOptions, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "@/redux/store";
import { I18nextProvider } from "react-i18next";
import i18n from "./i18n";
import { BrowserRouter } from "react-router-dom";
import { ReactElement } from "react";

interface AllTheProvidersProps {
    children: React.ReactNode;
}

const AllTheProviders: React.FC<AllTheProvidersProps> = ({ children }) => {
    return (
        <Provider store={store}>
            <I18nextProvider i18n={i18n}>
                <BrowserRouter>{children}</BrowserRouter>
            </I18nextProvider>
        </Provider>
    );
};

const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
