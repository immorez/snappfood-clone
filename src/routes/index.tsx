import { ReactElement, lazy } from "react";
import { Navigate, useRoutes } from "react-router-dom";
import Loadable from "./Loadable";
import MainLayout from "@/components/MainLayout";

const Home = Loadable(lazy(() => import("@/pages/home")));
const Restaurants = Loadable(lazy(() => import("@/pages/restaurants")));
const NotFound = Loadable(lazy(() => import("@/pages/404")));

/**
 * Router component that defines the application's routes using 'useRoutes'.
 * It includes a route for the Home component and a catch-all route for 404 errors.
 *
 * @returns {ReactElement} The router component.
 */

export default function Router(): ReactElement | null {
    return useRoutes(
        [
            {
                path: "/",
                element: <MainLayout />,
                children: [
                    {
                        path: "/",
                        index: true,
                        element: <Home />,
                    },
                    {
                        path: "/restaurants",
                        element: <Restaurants />,
                    },
                    {
                        path: "/404",
                        element: <NotFound />,
                    },
                ],
            },
            { path: "*", element: <Navigate to="/404" replace /> },
        ],
        location,
    );
}
