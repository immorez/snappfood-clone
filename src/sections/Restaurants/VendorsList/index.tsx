import { defaultLat, defaultLong, defaultPageSize } from "@/config/constants";
import { useGetVendorsListQuery } from "@/hooks/services/restaurantApiHooks";
import { DataEnumType, IVendor } from "@/redux/services/restaurantApi/types";
import React, { useMemo } from "react";
import TextDataItem from "./TextDataItem";
import VendorItem from "./VendorItem";

const VendorsList = function () {
    const { data: vendorsList } = useGetVendorsListQuery({
        lat: defaultLat,
        long: defaultLong,
        page_size: defaultPageSize,
        page: 0,
    });

    const sections = useMemo(() => {
        if (vendorsList) {
            return vendorsList.data.finalResult.map((item) => {
                switch (item.type) {
                    case DataEnumType.TEXT:
                        return <TextDataItem text={item.data as string} />;

                    case DataEnumType.VENDOR:
                        return <VendorItem {...(item.data as IVendor)} />;
                }
            });
        }
        return null;
    }, [vendorsList]);
    return (
        <div className="vendors-list">{React.Children.toArray(sections)}</div>
    );
};

export default React.memo(VendorsList);
