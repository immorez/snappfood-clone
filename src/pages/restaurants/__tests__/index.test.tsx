import { fireEvent, screen, waitFor } from "@testing-library/react";
import Restaurants from "..";
import { render } from "@/setupTests";
import { PATH_MAIN } from "@/routes/paths";

jest.mock("react-i18next", () => ({
    useTranslation: () => ({ t: (key: string) => key }),
}));

describe("Restaurants component", () => {
    it("renders Restaurants component correctly", async () => {
        const { container } = render(<Restaurants />);

        await waitFor(async () => {
            expect(screen.queryByTestId("skeleton")).toBeNull();
        });

        expect(container.firstChild).toMatchSnapshot();
    });

    it("clicking back button navigates to the home page", async () => {
        render(<Restaurants />, {
            routerOptions: { initialEntries: ["/restaurants"] },
        });
        const backButton = screen.getByRole("link", { name: "" });
        fireEvent.click(backButton);

        await waitFor(() => {
            expect(window.location.pathname).toBe(PATH_MAIN.home);
        });
    });
});
