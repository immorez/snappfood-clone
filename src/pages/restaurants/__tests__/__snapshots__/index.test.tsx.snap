// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Restaurants component renders Restaurants component correctly 1`] = `
<div
  class="restaurants"
>
  <a
    class="back-button"
    href="/"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 320 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l192 192c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L77.3 256 246.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-192 192z"
      />
    </svg>
  </a>
  <ul
    class="vendors-list"
  >
    <div
      class="skeleton"
    >
      <div
        class="skeleton--cover pulse"
      />
      <div
        class="skeleton--title pulse"
      />
      <div
        class="skeleton--description pulse"
      />
    </div>
    <div
      class="skeleton"
    >
      <div
        class="skeleton--cover pulse"
      />
      <div
        class="skeleton--title pulse"
      />
      <div
        class="skeleton--description pulse"
      />
    </div>
    <div
      class="skeleton"
    >
      <div
        class="skeleton--cover pulse"
      />
      <div
        class="skeleton--title pulse"
      />
      <div
        class="skeleton--description pulse"
      />
    </div>
  </ul>
</div>
`;
